### ISSUE TITLE
JSCS modifications to one-var

### ISSUE THREAD

**Issue by confuser** [eslint/eslint#4680]
Starting to migrate from JSCS/JSHint and I was looking for the equivalent of [requireMultipleVarDecl](http://jscs.info/rule/requireMultipleVarDecl)

Unfortunately `one-var` doesn't cover this, as it prevents multiple variable declarations in a single scope. I'm looking for allowing multiple variables throughout the scope, but preventing consecutive var declarations.

Invalid:

``` js
var a = 0
var b = 1
var c = 2

console.log('hello')

var d
var e
```

Valid:

``` js
var a = 0
  , b = 1
  , c = 2

console.log('hello')

var d, e
```
##

**Comment by eslintbot** [comment]
@confuser Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
It sounds like this would have to be an additional option for `one-var`, just not sure what the option name would be.

**Comment by confuser** [comment]
@nzakas what about setting the option to multi?

E.g.
```
"one-var": [2, {
    "var": "multi",
    "let": "multi",
    "const": "multi"
}]

"one-var": [2, "multi"]
```

**Comment by nzakas** [comment]
"Multi" is pretty ambiguous in this context because it's the opposite of one (and we already have "always" and "never")

**Comment by platinumazure** [comment]
What about something like "non-consecutive" or "non-consecutive-only"?

**Comment by SpadeAceman** [comment]
It sounds like the main request here is for a separate rule to allow or prevent multiple separate `var`/`let`/`const` statements in sequence. (`combine-seq-var-declarations` perhaps?)

This rule would enforce or prevent a single `var`/`let`/`const` statement for sequential variable declarations, regardless of the location of those `var`/`let`/`const` statements. (In other words, this seems like a rule that ought to be independent of `one-var`'s scope-related concerns.)

Valid when `never` specified (sequential variable declarations are never combined together):
```Javascript
var a = 0
var b = 1
var c = 2

console.log('hello')

var d
var e
```

Valid when `always` specified (sequential variable declarations are always combined together):
```Javascript
var a = 0
  , b = 1
  , c = 2

console.log('hello')

var d, e
```

**Comment by nzakas** [comment]
It can't be a separate rule because it will clash with `one-var`.

**Comment by confuser** [comment]
@nzakas Can this be added to the JSCS Compatibility milestone as part of #5856?

**Comment by nzakas** [comment]
We still need a concrete proposal for this rule.

**Comment by zxqfox** [comment]
Repeating here to close #5910 as dup:

[requireMultipleVarDecl](http://jscs.info/rule/requireMultipleVarDecl), [disallowMultipleVarDecl](http://jscs.info/rule/disallowMultipleVarDecl) → http://eslint.org/docs/rules/one-var (?)

`onevar` doesn't cover atm these cases:
- `{requireMultipleVarDecl: true}`
- `{requireMultipleVarDecl: {allExcept:["require"]}}`
- `{disallowMultipleVarDecl: {allExcept:["require"]}}`
- `{disallowMultipleVarDecl: {allExcept:["undefined"]}}` (if it's not the same as `{var: {uninitialized: "always"}}`)

Current `one-var` options: `{ var | let | const | uninitialized | initialized: "always"|"never" }`.

**Comment by mysticatea** [comment]
Now, I have a question.

```js
var a = 0
  , b = 1
  , c = 2

console.log('hello')

var d, e  // Why 'one-var' rule is warning this declaration?
```

I feel it's the responsibility of `vars-on-top` rule.

**Comment by ilyavolodin** [comment]
@mysticatea I think it's responsibility of both rules. `one-var` should warn you that there are multiple `var` keywords used, and `vars-on-top` should warn that variable declaration is not at the begging of the block.
I think your screenshot actually looks correct.

**Comment by mysticatea** [comment]
Doesn't `one-var` which warns only neighbor declaration work fine?
`vars-on-top` indicates developers to move declarations to the neighbor of the first declaration.

**Comment by ilyavolodin** [comment]
@mysticatea `one-var` in it's default configuration basically requires you to only have 1 `var` keyword per file/function/class. `vars-on-top` doesn't care how many `var` keywords you have, as long as they all are located at the begging of the block.

**Comment by mysticatea** [comment]
My question is, why `one-var` is "per file/function"? It sounds fine that a check of only neighbor declarations is.

**Comment by mysticatea** [comment]
My proposal is 

- Adds `"neighbor"` option value.
- Adds 3rd optional object `{"require": value}` to each pattern.

This avoids breaking changes.

# one-var

```js
{
    "one-var": [
        "error", 
        "always" or "never" or "neighbor",
        {
            "require": "always" or "never" or "neighbor"
        }
    ],
    // or
    "one-var": [
        "error", 
        {
            "initialized": "always" or "never" or "neighbor",
            "uninitialized": "always" or "never" or "neighbor"
        },
        {
            "require": "always" or "never" or "neighbor"
        }
    ],
    // or
    "one-var": [
        "error", 
        {
            "var": "always" or "never" or "neighbor",
            "let": "always" or "never" or "neighbor",
            "const": "always" or "never" or "neighbor"
        },
        {
            "require": "always" or "never" or "neighbor"
        }
    ]
}
```

- `"always"` (default) - keep the current behavior as is.
- `"never"` - keep the current behavior as is.
- `"neighbor"` - `one-var` warns neighbor variable declarations. This is the same behavior as JSCS rules.
- `require` (default is no overriding) - This is the setting for `retuire()`. JSCS rules have this option.

----

```js
/*eslint one-var:["error", "neighbor"]*/

//----------------------------------------------------------------
// Good
let a = 0,
    b = 1;

foo();

let c = 2;

//----------------------------------------------------------------
// Bad
let a = 0,
    b = 1;
let c = 2;
```


```js
/*eslint one-var:["error", "never", {"require": "always"}]*/

//----------------------------------------------------------------
// Good
let a = require("a"),
    b = require("a");

let c = 2;
let d = 3;

//----------------------------------------------------------------
// Bad
let a = 0,
    b = 1;
let c = require("c");
let d = require("d");
```

**Comment by ilyavolodin** [comment]
@mysticatea Basically `one-var` with always configuration says "always combine all of the variable declarations into a single declaration". And never is the reverse of that. Having a single `var` keyword per file/function is a pretty common pattern. We could add those two properties for JSCS compatibility, but the default configuration should stay the same. I don't think there's an overlap with `vars-on-top` at all. `one-var` controls number of `var` keywords, and `vars-on-top` controls location of variable declarations. You could enable `one-var` and put all of the variable declarations at the bottom of the function, for example, and `one-var` would be perfectly fine.

**Comment by mysticatea** [comment]
hmmm, I think the pretty common pattern is a combination of `one-var` and `vars-on-top`. not `one-var`.

**Comment by mysticatea** [comment]
Anyway, I don't like breaking changes.
So my proposal here is above.

**Comment by nzakas** [comment]
@ilyavolodin are you saying "neighbor" is unnecessary because `vars-on-top` covers that case?

**Comment by ilyavolodin** [comment]
No, as far as I understand the proposal neighbor option actually doesn't have anything to do with `vars-on-top` at all. I assume what it would do is force combining blocks of variable declarations into a single one. None of the blocks have to be on top. I was just commenting on the example that @mysticatea posted above. I think he was thinking that `one-var` and `vars-on-top` are conflicting, and I was pointing out that they are not, and both are working correctly.

**Comment by nzakas** [comment]
Ah I see, OK. 

Then lets refocus on the current proposal with "neighbor". I don't think that's the clearest term we could use, though the option seems correct. Other thoughts?

Also, I'm a bit uneasy with an option for `require` being built into this rule. In general, we have tried to keep environment-specific options out of generic rules.

**Comment by mysticatea** [comment]
I'm sorry for I had gone far.

Please reword the `"neighbor"`. I just took it from a Japanese‐English dictionary. :smile: 

Indeed, the 2nd option object is awkward, so I hope better idea caming.
I think `require()` is used generally in Node and front-end (browserify, webpack, ...). And I understand the request to separate incoming things from internal things.
Hmm, is actual requirement #6175, not the `require` option of my proposal?

**Comment by andrewvaughan** [comment]
The `neighbor` and #6175 suggestions are different.  The neighbor suggestions recommends that one-var accept situations in which there are non-declarative statements between variable blocks.  The recommendation from #6175 is an ask for compliance with the `no-mixed-requires` rule that is not compatible with `one-var`

**Comment by nzakas** [comment]
Maybe `consecutive ` instead of `neighbor `?

**Comment by mysticatea** [comment]
`consecutive` sounds pretty good.

I looked [the document of `no-mixed-requires`](https://github.com/eslint/eslint/blob/master/docs/rules/no-mixed-requires.md), yeah, `no-mixed-requires` is conflicting with `one-var: always`.
Hmm, I felt that adding the same behavior as  the `grouping` option of `no-mixed-requires` rule is overspec for `one-var`...

**Comment by nzakas** [comment]
Let's keep this focused on `consecutive` and open a new issue for `no-mixed-requires`.

I'm 👍

**Comment by mysticatea** [comment]
:+1:

**Comment by ilyavolodin** [comment]
👍  for `consecutive`

**Comment by nzakas** [comment]
@eslint/eslint-team we need one more 👍 here

**Comment by alberto** [comment]
Here is another 👍  , and an alternative name suggestion just in case: `grouped`

**Comment by platinumazure** [comment]
@rkaneko I see you've written a commit for this in July. Is this something you might be willing to submit as a pull request?

**Comment by kaicataldo** [comment]
I'm working on this!
